improve versioning docs (#721)
ubuntu/ubuntu-project-docs: docs/how-ubuntu-is-made/concepts/version-strings.md

  revision is `3`.
- Ubuntu has uploaded **two** revisions to Debian `9.7-3`, so the Ubuntu
  revision is `ubuntu2`.

Since Ubuntu's versioning specifics are derived from Debian, it is worthwhile to
also review the
[Debian control field "Version"](https://www.debian.org/doc/debian-policy/ch-controlfields.html#version)
to understand the foundational principles.

(comparing-versions)=
## Comparing versions

To compare two version strings, use {command}`dpkg --compare-versions`. The
available comparison operators are `lt`, `le`, `eq`, `ne`, `ge`, and `gt`:

```{code}
:class: codeblock-wrap

$ dpkg --compare-versions 1.2.3-3ubuntu1 lt 1.2.4-2ubuntu2 && echo true || echo false
```

It is worth noting the special sorting rules regarding the `~` character.
Namely, `~` sorts _before_ everything else, including the empty string. So `1.0~rc1` sorts _before_ `1.0`.

(standard-version-bumps)=
## Standard version bumps

The great majority of uploads fall into one of three routine cases: making an
Ubuntu change in the development release, merging a newer Debian version, or
syncing an unmodified Debian version. These are covered first; the special cases

> follow.

Reviewer: ```suggestion
The great majority of uploads fall into one of three routine cases:

- making an Ubuntu change in the development release
- merging a newer Debian version
- syncing an unmodified Debian version

These are covered first; the special cases follow.
```

> (version-devel-upload)=

Reviewer: ```suggestion
(uploading-to-the-development-release)=
```

### Uploading to the development release

Changes within the development release are made by adding or incrementing the
number immediately following the `ubuntu` string.

The first Ubuntu modification adds `ubuntu1` to the Debian revision, and
subsequent Ubuntu changes increment that suffix (`ubuntu1` → `ubuntu2`).

> Example: adding a change to `2.0-2` in the development release produces
  `2.0-2ubuntu1`. A further change produces `2.0-2ubuntu2`.

If the current version instead ends in a `buildN`
({ref}`no-change rebuild <version-no-change-rebuilds>`) suffix, the package is
considered unchanged from the Debian revision, so a new Ubuntu change starts the
counter afresh at `ubuntu1` (for example `2.0-3build2` → `2.0-3ubuntu1`).

> (version-merging-from-debian)=

Reviewer: ```suggestion
(merging-from-debian)=
```

### Merging from Debian

If Ubuntu carries delta, the package cannot be

> {ref}`automatically synced <version-syncing-from-debian>`, since that delta

Reviewer: ```suggestion
If Ubuntu carries delta, the package cannot be
{ref}`automatically synced <synching-from-debian>`, since that delta
```

would be lost. Instead, an Ubuntu developer periodically
{ref}`merges <merge-process>` the Ubuntu delta with the newer content from
Debian (and, indirectly, from upstream).

On a merge, the version becomes the new Debian version with the Ubuntu counter
reset to `ubuntu1`.

> Example: merging Debian `3.1-2` into the development release while retaining
  former Ubuntu delta from `2.1-1ubuntu2` produces `3.1-2ubuntu1`.

> (version-syncing-from-debian)=

Reviewer: ```suggestion
(synching-from-debian)=
```

> ### Syncing from Debian

Reviewer: ```suggestion
### Synching from Debian
```

or _syncs_, new package versions from Debian. Only packages with no Ubuntu delta
(determined via the presence of the `ubuntu` string in the Ubuntu revision) are synced
automatically.

When a sync happens -- automatically, or as a
{ref}`manual sync <how-to-request-a-sync>` -- the version is **identical to the
Debian version**; no suffix is added.

> Example: syncing Debian `2.0-3` on top of Ubuntu `2.0-2` or `2.0-2build1`
  produces `2.0-3`.

There are two common cases that require a manual sync:

- After the development release reaches {ref}`debian-import-freeze`, the sync
  automation is disabled, so any required update must be
  {ref}`manually synced <how-to-request-a-sync>`.
- When a package had Ubuntu delta that is no longer needed with the new Debian
  version (for example, Debian adopted the Ubuntu changes, or they were
  cherry-picked patches now present upstream), a
  {ref}`manual re-sync <how-to-request-a-sync>` drops the delta. The version is
  again identical to Debian.

(version-debian-source-suffixes)=
### Debian source-repackaging suffixes (`+dfsg`, `+ds`)

You will often see extra markers on the **upstream-version** portion of a

> package (before the `-debian_revision`), most commonly `+dfsg` and `+ds`. Despite

Reviewer: ```suggestion
Extra markers on the **upstream-version** portion of a
package (before the `-debian_revision`) are often used, most commonly `+dfsg` and `+ds`. Despite
```

being added by Debian, they are part of the upstream version, not the Debian
revision. They indicate that the source tarball Debian ships differs from the
pristine upstream release. Ubuntu inherits them unchanged: there is never a need
to add, remove, or renumber them ourselves.

- **`+dfsg`** (Debian Free Software Guidelines) -- the upstream tarball was
  repackaged to remove content that does not meet the
  [DFSG](https://www.debian.org/social_contract#guidelines), such as non-free
  files. For example, `7.80+dfsg1-5`.
- **`+ds`** (Debian source) -- the tarball was repackaged for reasons other than
  licensing, for example to drop bundled third-party code, pre-built artifacts,
  or very large test data. For example, `1.2.3+ds-1`.

A trailing number (`+dfsg1`, `+ds2`) counts the repackaging revisions. Because
`+` sorts _after_ the bare version, `7.80+dfsg1` is treated as a _later_ version
than a hypothetical pristine `7.80`, which keeps upgrade ordering sensible.

(special-cases-ubuntu-revision)=

> ## Special cases: the Ubuntu revision

Reviewer: ```suggestion
## Special cases: Ubuntu revision
```

If a package in Ubuntu has differences that prevent it from being automatically

> updated to the latest Debian version, we say the package has **Ubuntu delta**

Reviewer: ```suggestion
A package in Ubuntu with differences that prevent it from being automatically
updated to the latest Debian version has **Ubuntu delta**
```

(or just "delta"). Most Ubuntu-specific changes constitute delta, and the
`ubuntuN` number is incremented with each such change.

If the Ubuntu revision does _not_ begin with `ubuntu`, the package **will** be

> automatically synced and overwritten by a later Debian upload. The following

Reviewer: ```suggestion
Unless the Ubuntu revision begins with `ubuntu`, the package **is** automatically synced and overwritten by a later Debian upload. The following
```

sections cover the cases where a different Ubuntu revision string is used.

> (version-no-change-rebuilds)=

Reviewer: ```suggestion
(no-change-rebuilds-build)=
```

### No-change rebuilds (`build`)

Sometimes a package needs its binaries rebuilt without any source change -- for
example, to re-link against a newer {term}`ABI` of a dependency as part of a
{ref}`transition <transitions>`.

A rebuild should not prevent future syncing, so in this case:

- If the current version has no delta, add a `buildN` suffix rather than
  `ubuntuN` (`2.0-2` → `2.0-2build1`).
- If a `buildN` suffix is already present, increment it
  (`2.0-2build1` → `2.0-2build2`).
- If an `ubuntuN` suffix is already present, there would be no auto-sync anyway,

>   so simply increment it (`2.0-2ubuntu2` → `2.0-2ubuntu3`).

Reviewer: ```suggestion
Current version has no delta
: Add a `buildN` suffix rather than `ubuntuN` (`2.0-2` → `2.0-2build1`).

`buildN` suffix is already present
: Increment it (`2.0-2build1` → `2.0-2build2`).
 
`ubuntuN` suffix is already present (there would be no auto-sync anyway)
: Increment it (`2.0-2ubuntu2` → `2.0-2ubuntu3`).
```

For {ref}`native packages <version-native-packages>`, it is weakly defined
whether a no-change rebuild is a version increment or a `build` suffix; both are
present in the Archive and both work. Follow whatever the package has used so
far (for example `2.0ubuntu.build1`).

> (version-syncable-changes)=

Reviewer: ```suggestion
(syncable-changes-maysync-willsync)=
```

`maysync` and `willsync`.

Both indicate that the package may still be automatically synced with Debian
(only an `ubuntu` revision prevents syncing). The only difference is ordering:
`maysync1` sorts _before_ `ubuntu1`, while `willsync1` sorts _after_ `ubuntu1`.
Therefore:

- Use `maysync` if the package previously had **no** delta, so that delta can
  still be added later in the standard way (`ubuntu` > `maysync`).
- Use `willsync` if the package previously **had** delta, to indicate that it is
  now syncable again (`willsync` > `ubuntu`), so the next Debian update
  overwrites it automatically.

> Example: A syncable upload on top of `2.0-1` is `2.0-1maysync1`, and a syncable upload on top of `2.0-1ubuntu1` is `2.0-1willsync1`.

It may seem strange to upload a syncable version on top of a version that is
_already_ synced with Debian, but it can be necessary. For example:

- You need a change present only in an unreleased Debian version.
- You are working around a temporary packaging issue (for example, a
  dependency-related build failure) with a temporary delta (`maysync1`) that is
  then immediately dropped (`maysync2`).

> (version-srus)=

Reviewer: ```suggestion
(version-stable-release-updates-srus)=
```

### Stable release updates (SRUs)

After a version of Ubuntu is released, changes follow a slightly different
scheme that guarantees upgradability to later releases. Only the _Ubuntu
revision_ changes:

- Increment `Y` in the numeric `ubuntuX.Y` suffix (`ubuntu3.1` → `ubuntu3.2`).
- Never increment `X`; an Ubuntu delta is always captured in `Y`.
- If this is the first change via the SRU process, add `.1`
  (`ubuntu3` → `ubuntu3.1`).
- If there was no `ubuntuX` before, use `ubuntu0.1`. The `0` records that there
  was no Ubuntu delta before this first SRU change (`2.0-2` → `2.0-2ubuntu0.1`).

```{important}
A given version string may only ever be uploaded **once** to the Archive, across
_all_ releases. If the same fix is applied to a package that has the same version
in several releases, using the same version string in each would create a
conflict and break upgradability. To disambiguate, insert the per-release
`YY.MM` version between `ubuntuX` and the `.Y` increment.
```

> Example: applying the same fix to `2.0-2` in both 25.04 and 24.04 produces
  `2.0-2ubuntu0.25.04.1` and `2.0-2ubuntu0.24.04.1` respectively.

> (version-backport-from-upstream)=

Reviewer: ```suggestion
(backporting-a-new-upstream-version)=
```

### Backporting from the development release

If instead the backport has no _meaningful differences_ from what is in the
current development release (a common practice for packages that keep the same
version everywhere), with only minimal backporting adaptations, it should:

- Start with the **same version** as the one in the development release, to
  signal that it is basically identical but backported.
- Add a `~` so it sorts _earlier_ than the version in the development release.
- Add a per-release `YY.MM` suffix.
- Add a `.1` increment for subsequent per-release SRU uploads.

> Example: uploading `3.1`, more or less identical to the `3.1-1ubuntu2` in the
  development release, targeting 22.04 produces `3.1-1ubuntu2~22.04.1`.

As with the {ref}`backport from upstream <version-backport-from-upstream>`, the new version is independent of the version already present in the target release.

```{note}
If the package in the development release is a
{ref}`native package <version-native-packages>`, use _the same version_ as in
the development release, with the `~YY.MM.1` suffix (for example a devel version
of `3.1` becomes `3.1~22.04.1`).
```

(special-cases-debian-revision)=

> ## Special cases: the Debian revision

Reviewer: ```suggestion
## Special cases: Debian revision
```

> (version-ahead-of-debian)=

Reviewer: ```suggestion
(going-ahead-of-debian)=
```

### Going ahead of Debian

Sometimes Ubuntu packages a version that Debian has never packaged -- for
example, a newer upstream release, or a specific upstream commit. In that case
we use a Debian revision of `0` to indicate that this is not based on any Debian
packaging version.

Because such a version is our own change, it must always carry an Ubuntu
revision as well, giving `-0ubuntu1` for the first development upload (which also
serves as the increment counter for further uploads).

> Example: uploading upstream `3.1` while Debian is not yet at `3.1` produces
  `3.1-0ubuntu1`.

The `-0` marker is specific to the Debian revision. The way the _upstream_
portion is written for pre-releases and git snapshots is described in
{ref}`special-cases-upstream-version` below -- note that Debian could package the
very same upstream snapshot, in which case it would use its own Debian revision
(such as `-1`) rather than `-0`.

(special-cases-upstream-version)=

> ## Special cases: the upstream version

Reviewer: ```suggestion
## Special cases: upstream version
```

The following conventions shape the **upstream version** portion only. When the
snapshot is one Ubuntu has packaged ahead of Debian, combine them with the
`-0ubuntuN` revision from {ref}`version-ahead-of-debian`.

> (version-tagged-pre-releases)=

Reviewer: ```suggestion
(tagged-pre-releases)=
```

usually backported with the same content to all active releases.

That combination means some of the normal rules do not apply, blending the
_native package_ and _backport from upstream_ cases. For such a package:

- The first element of the version matches the upstream version it represents
  (for example `2.67`).
- Because it is backported to multiple releases at the same version, but being
  native cannot use a `-`, the per-release suffix is added with `+ubuntuYY.MM`.
  (Former versions used just `+YY.MM`, but `+ubuntuYY.MM` is preferred for
  symmetry and to avoid auto-sync conflicts.)
- If further iterations are needed for the same upstream version and the same
  target release, add a `.n` increment.

| Previous                    | New upstream 2.67  | Change within 2.66     |
| --------------------------- | ------------------ | ---------------------- |
| LTS: `2.66+ubuntu24.04`     | `2.67+ubuntu24.04` | `2.66+ubuntu24.04.1`   |
| Devel: `2.66+ubuntu25.04`   | `2.67+ubuntu25.04` | `2.66+ubuntu25.04.1`   |
| LTS: `2.66+ubuntu24.04.1`   | `2.67+ubuntu24.04` | `2.66+ubuntu24.04.2`   |
| Devel: `2.66+ubuntu25.04.1` | `2.67+ubuntu25.04` | `2.66+ubuntu25.04.2`   |

> (version-rolling-back)=

Reviewer: ```suggestion
(rolling-back)=
```

| `2.0-2ubuntu2` | `3.1-2ubuntu1` | `3.1+really2.0-2ubuntu2`                  | Devel         |
| `7.80+dfsg1-5` | `7.91+dfsg1-1` | `7.91+dfsg1+really7.80+dfsg1-1ubuntu1`    | Ubuntu upload |
| `7.80+dfsg1-5` | `7.91+dfsg1-1` | `7.91+dfsg1+really7.80+dfsg1-1`           | Debian upload |
| `7.80+dfsg1-5` | `7.91+dfsg1-1` | `7.91+dfsg1+really7.80+dfsg1-1ubuntu0.1`  | Ubuntu SRU    |

(version-epochs)=
### Epochs

In extraordinarily rare circumstances, the versioning may need a complete reset.
This can happen because upstream radically altered their scheme or to resolve a serious
mistake. The version format allows an **epoch** to be prepended for this.

The epoch is a single integer followed by a colon at the beginning of the version
string. If no epoch is present (as for most packages), it is assumed to be `0`.

Changing the epoch should be done with care and requires consensus in Debian.
Epochs are explicitly **not** used when a package needs to be reverted to a
previous version (see {ref}`version-rolling-back`). See also
[epochs should be used sparingly](https://www.debian.org/doc/debian-policy/ch-controlfields.html#epochs-should-be-used-sparingly).

> (versioning-examples)=

Reviewer: ```suggestion
(version-examples)=
```

## Examples

> (version-timeline-example)=

Reviewer: ```suggestion
(timeline-example-standard-flow)=
```

### Timeline example: standard flow

The following pathway illustrates the standard flow of a package starting with a
new upstream release. Suppose XX.04 is the current development release:

1. The current version of {pkg}`libfoo` in both Debian and Ubuntu is `6.6-1`.
2. The {pkg}`libfoo` maintainers release a new upstream version, `6.7`.
3. Debian packages it as `6.7-1`.
4. It syncs into Ubuntu, still as `6.7-1`.
5. Ubuntu rebuilds the package and uploads it as `6.7-1build1`.
6. Ubuntu adds some delta, uploading it as `6.7-1ubuntu1`.
7. Debian makes a change, uploading it as `6.7-2`.
8. Ubuntu rebases its delta onto the new Debian version, uploading `6.7-2ubuntu1`.
9. Ubuntu makes another patch, uploading `6.7-2ubuntu2`.
10. XX.04 is released.
11. A bug is found that needs patching in both XX.04 and XX.10:
    - In XX.10 (the new development release), the fix is uploaded as
      `6.7-2ubuntu3`.

>     - The same fix is SRUed into XX.04 as `6.7-2ubuntu2.1`.

Reviewer: ```suggestion
    - The same fix is release as SRU into XX.04 as `6.7-2ubuntu2.1`.
```

> (version-timeline-ahead-of-debian)=

Reviewer: ```suggestion
(timeline-example-going-ahead-of-debian)=
```

### Timeline example: Going Ahead of Debian

1. {pkg}`libbar` is `2.0-1` in both Debian and Ubuntu.
2. Upstream releases `2.1`.
3. Ubuntu wants the new upstream release before Debian is ready, so it {ref}`merges from upstream <version-ahead-of-debian>`, uploading `2.1-0ubuntu1`.
4. Ubuntu makes a patch, uploading `2.1-0ubuntu2`, and submits their patch to upstream.
5. Ubuntu's patch is accepted upstream.
6. Upstream releases version `2.2`, which includes Ubuntu's patch and a different fix.
7. Ubuntu backports `2.2`'s other fix to `2.1`. Since all delta will be gone in `2.2`, Ubuntu uploads this fix as `2.1-0willsync1`
8. Debian packages `2.2` as `2.2-1`.
9. Ubuntu automatically syncs `2.2-1` from Debian. Both distributions are in sync once more.

> (version-timeline-rollback)=

Reviewer: ```suggestion
(timeline-example-rolling-back)=
```

| `2ubuntu1` (native Debian, delta)  | Native devel upload             | —                               | `2ubuntu2`                              |
| `2.0ubuntu` (native Ubuntu)        | Native devel upload             | —                               | `2.1ubuntu` or `3.0ubuntu`              |
| `2.0ubuntu` (native Ubuntu)        | Native SRU upload               | —                               | `2.0ubuntu0.1`                          |
| `2.66+ubuntu24.04`                 | Almost-native upload            | Upstream `2.67`                 | `2.67+ubuntu24.04`                      |
| `2.66+ubuntu24.04`                 | Almost-native upload (in-place) | —                               | `2.66+ubuntu24.04.1`                    |
| `3.1-2ubuntu1`                     | Rollback (devel)                | Target `2.0-2ubuntu2`           | `3.1+really2.0-2ubuntu2`                |
| `7.91+dfsg1-1`                     | Rollback (Ubuntu upload)        | Target `7.80+dfsg1`             | `7.91+dfsg1+really7.80+dfsg1-1ubuntu1`  |
| `7.91+dfsg1-1`                     | Rollback (Ubuntu SRU)           | Target `7.80+dfsg1`             | `7.91+dfsg1+really7.80+dfsg1-1ubuntu0.1`|

> (version-flowchart)=

Reviewer: ```suggestion
(choosing-a-version-string)=
```

## Choosing a version string

The following flowchart summarizes the decisions above and points to the
matching version string for your upload.

```{mermaid}
%% mermaid flowcharts documentation: https://mermaid.js.org/syntax/flowchart.html
%%{ init: { "flowchart": { "curve": "linear", "nodeSpacing": 35, "rankSpacing": 55, "htmlLabels": true } } }%%
flowchart LR
    classDef leaf fill:#E6F4EA,stroke:#34A853,color:#000
    classDef q fill:#FEF7E0,stroke:#F9AB00,color:#000
    classDef special fill:#FCE8E6,stroke:#EA4335,color:#000

    Start(["<b>Which version<br>string do I pick?</b>"])
    Start --> QWhat{"What are you<br>uploading?"}:::q

    QWhat -->|"A new <code>devel</code><br>release"| QDev{"Source of the<br>change?"}:::q
    QWhat -->|"A Stable Release<br>Update (SRU)"| QSru{"Prior Ubuntu delta<br>in this release?"}:::q
    QWhat -->|"A Backport to a<br>stable release"| QBack{"Backport content?"}:::q

>     QWhat -->|"A <b>rollback</b> (to a<br>lower version)"| QRollW{"In devel or<br>an SRU?"}:::q

Reviewer: ```suggestion
    QWhat -->|"New <code>devel</code><br>release"| QDev{"Source of the<br>change?"}:::q
    QWhat -->|"Stable Release<br>Update (SRU)"| QSru{"Prior Ubuntu delta<br>in this release?"}:::q
    QWhat -->|"Backport to a<br>stable release"| QBack{"Backport content?"}:::q
    QWhat -->|"<b>Rollback</b> (to a<br>lower version)"| QRollW{"In devel or<br>an SRU?"}:::q
```
